src/printer/tvmscript_unified_printer.cc

```diff
@@ -0,0 +1,738 @@
+    auto size = doc->elements.size();
+    if (size == 0) {
+      output_ << "tuple()";
+    } else if (size == 1) {
+      output_ << "(";
+      PrintDoc(doc->elements[0]);
+      output_ << ",)";
+    } else {
+      PrintJoinedElements("(", doc->elements, ", ", ")");
+    }
+  };
+
+  virtual void PrintDoc(const SeqStmtDoc& doc) override {
+    for (auto& stmt : doc->seq) {
+      PrintDoc(stmt);
+    }
+  };
+
+  virtual void PrintDoc(const ScopeDoc& doc) override {
+    NewLine() << "with ";
+    PrintDoc(doc->scope);
+    output_ << ":";
+
+    PrintWithIncreasedIndent(doc->body);
+  };
+
+  virtual void PrintDoc(const ForDoc& doc) override {
+    NewLine() << "for ";
+    PrintDoc(doc->target);
+    output_ << " in ";
+    PrintDoc(doc->iter);
+    output_ << ":";
+
+    PrintWithIncreasedIndent(doc->body);
+  };
+
+  virtual void PrintDoc(const AssignDoc& doc) override {
+    // TODO: Assign Kind, Type Annotation
+    NewLine();
+    PrintDoc(doc->target);
+    if (doc->type) {
+      output_ << ": ";
+      PrintDoc(doc->type.value());
+    }
+    if (doc->value) {
+      output_ << " = ";
+      PrintDoc(doc->value.value());
+    }
+  };
+
+  virtual void PrintDoc(const ExprTypeDoc& doc) override { PrintDoc(doc->expr); };
+
+  virtual void PrintDoc(const TypeCallDoc& doc) override {
+    PrintDoc(doc->base);
+    PrintJoinedElements("[", doc->args, ", ", "]");
+  };
+
+  virtual void PrintDoc(const FunctionDoc& doc) override {
+    NewLine() << "@" << tir_prefix_ << ".prim_func";
+    NewLine() << "def ";
+    PrintDoc(doc->name);
+    PrintJoinedElements("(", doc->args, ", ", ")");
+    output_ << " -> ";
+    PrintDoc(doc->return_type);
+    output_ << ":";
+    PrintWithIncreasedIndent(doc->body);
+  };
+
+  virtual void PrintDoc(const FunctionArgDoc& doc) override {
+    PrintDoc(doc->name);
+    output_ << ": ";
+    PrintDoc(doc->type);
+  };
+
+ private:
+  String tir_prefix_;
+
+  int indent_spaces_ = 4;
+
+  void IncreaseIndent() { indent_ += indent_spaces_; }
+
+  void DecreaseIndent() { indent_ -= indent_spaces_; }
+
+  void PrintWithIncreasedIndent(const Doc& doc) {
+    IncreaseIndent();
+    PrintDoc(doc);
+    DecreaseIndent();
+  }
+
+  template <typename ContainerType>
+  void PrintJoinedElements(const std::string& left, const ContainerType& elements,
+                           const std::string& separator, const std::string& right,
+                           const std::string& default_content = "") {
+    output_ << left;
+    bool is_first = true;
+    for (auto& element : elements) {
+      if (is_first) {
+        is_first = false;
+      } else {
+        output_ << separator;
+      }
+      PrintObject(element);
+    }
+    if (is_first) {
+      // This means that `elements` is empty
+      output_ << default_content;
+    }
+    output_ << right;
+  }
+
+  template <typename... Arg>
+  void PrintTIRPrimitiveCall(const std::string& primitive, Arg... args) {
```

Comment:
Let's avoid using templates if possible

```suggestion
  void PrintTIRPrimitiveCall(const std::string& primitive, const std::vector<ExprDoc>& args) {
```